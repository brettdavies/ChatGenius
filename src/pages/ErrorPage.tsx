import { useRouteError, isRouteErrorResponse } from 'react-router-dom';

const ErrorPage = () => {
  const error = useRouteError();
  
  let errorMessage: string;
  
  if (isRouteErrorResponse(error)) {
    errorMessage = error.statusText || error.data;
  } else if (error instanceof Error) {
    errorMessage = error.message;
  } else if (typeof error === 'string') {
    errorMessage = error;
  } else {
    errorMessage = 'Unknown error';
  }

  return (
    <div className="min-h-screen flex items-center justify-center">
      <div className="text-center">
        <h1 className="text-4xl font-bold text-gray-900 mb-4">Oops!</h1>
        <p className="text-lg text-gray-600 mb-8">
          Sorry, an unexpected error has occurred.
        </p>
        <p className="text-gray-500">
          <i>{errorMessage}</i>
        </p>
      </div>
    </div>
  );
};

export default ErrorPage; 