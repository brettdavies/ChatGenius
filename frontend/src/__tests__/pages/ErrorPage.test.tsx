import { render, screen } from '@testing-library/react';
import { useRouteError } from 'react-router-dom';
import ErrorPage from '../../pages/ErrorPage';

// Mock useRouteError
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useRouteError: jest.fn()
}));

const mockUseRouteError = useRouteError as jest.Mock;

describe('ErrorPage', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('snapshots', () => {
    it('matches snapshot with Error object', () => {
      mockUseRouteError.mockReturnValue(new Error('Test error message'));
      const { container } = render(<ErrorPage />);
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot with route error response', () => {
      mockUseRouteError.mockReturnValue({
        statusText: 'Not Found',
        data: 'Page not found',
        status: 404,
        internal: true,
        error: new Error(),
      });
      const { container } = render(<ErrorPage />);
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot with string error', () => {
      mockUseRouteError.mockReturnValue('Direct string error message');
      const { container } = render(<ErrorPage />);
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot with unknown error type', () => {
      mockUseRouteError.mockReturnValue({ custom: 'error' });
      const { container } = render(<ErrorPage />);
      expect(container).toMatchSnapshot();
    });
  });

  it('renders common error elements', () => {
    mockUseRouteError.mockReturnValue(new Error('Test error'));

    render(<ErrorPage />);

    expect(screen.getByText('Oops!')).toBeInTheDocument();
    expect(screen.getByText('Sorry, an unexpected error has occurred.')).toBeInTheDocument();
  });

  it('displays error message from Error object', () => {
    const errorMessage = 'Test error message';
    mockUseRouteError.mockReturnValue(new Error(errorMessage));

    render(<ErrorPage />);

    expect(screen.getByText(errorMessage)).toBeInTheDocument();
  });

  it('displays error message from route error response with statusText', () => {
    const errorStatusText = 'Not Found';
    mockUseRouteError.mockReturnValue({
      statusText: errorStatusText,
      data: 'Some data',
      status: 404,
      internal: true,
      error: new Error(),
    });

    render(<ErrorPage />);

    expect(screen.getByText(errorStatusText)).toBeInTheDocument();
  });

  it('displays error message from route error response with data when statusText is empty', () => {
    const errorData = 'Error data message';
    mockUseRouteError.mockReturnValue({
      statusText: '',
      data: errorData,
      status: 500,
      internal: true,
      error: new Error(),
    });

    render(<ErrorPage />);

    expect(screen.getByText(errorData)).toBeInTheDocument();
  });

  it('displays string error message directly', () => {
    const errorMessage = 'Direct string error message';
    mockUseRouteError.mockReturnValue(errorMessage);

    render(<ErrorPage />);

    expect(screen.getByText(errorMessage)).toBeInTheDocument();
  });

  it('displays "Unknown error" for unhandled error types', () => {
    mockUseRouteError.mockReturnValue({ custom: 'error' });

    render(<ErrorPage />);

    expect(screen.getByText('Unknown error')).toBeInTheDocument();
  });

  it('handles null error gracefully', () => {
    mockUseRouteError.mockReturnValue(null);

    render(<ErrorPage />);

    expect(screen.getByText('Unknown error')).toBeInTheDocument();
  });

  it('handles undefined error gracefully', () => {
    mockUseRouteError.mockReturnValue(undefined);

    render(<ErrorPage />);

    expect(screen.getByText('Unknown error')).toBeInTheDocument();
  });
}); 