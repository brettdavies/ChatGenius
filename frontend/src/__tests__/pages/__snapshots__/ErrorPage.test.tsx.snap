// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorPage snapshots matches snapshot with Error object 1`] = `
<div>
  <div
    class="min-h-screen flex items-center justify-center"
  >
    <div
      class="text-center"
    >
      <h1
        class="text-4xl font-bold text-gray-900 mb-4"
      >
        Oops!
      </h1>
      <p
        class="text-lg text-gray-600 mb-8"
      >
        Sorry, an unexpected error has occurred.
      </p>
      <p
        class="text-gray-500"
      >
        <i>
          Test error message
        </i>
      </p>
    </div>
  </div>
</div>
`;

exports[`ErrorPage snapshots matches snapshot with route error response 1`] = `
<div>
  <div
    class="min-h-screen flex items-center justify-center"
  >
    <div
      class="text-center"
    >
      <h1
        class="text-4xl font-bold text-gray-900 mb-4"
      >
        Oops!
      </h1>
      <p
        class="text-lg text-gray-600 mb-8"
      >
        Sorry, an unexpected error has occurred.
      </p>
      <p
        class="text-gray-500"
      >
        <i>
          Not Found
        </i>
      </p>
    </div>
  </div>
</div>
`;

exports[`ErrorPage snapshots matches snapshot with string error 1`] = `
<div>
  <div
    class="min-h-screen flex items-center justify-center"
  >
    <div
      class="text-center"
    >
      <h1
        class="text-4xl font-bold text-gray-900 mb-4"
      >
        Oops!
      </h1>
      <p
        class="text-lg text-gray-600 mb-8"
      >
        Sorry, an unexpected error has occurred.
      </p>
      <p
        class="text-gray-500"
      >
        <i>
          Direct string error message
        </i>
      </p>
    </div>
  </div>
</div>
`;

exports[`ErrorPage snapshots matches snapshot with unknown error type 1`] = `
<div>
  <div
    class="min-h-screen flex items-center justify-center"
  >
    <div
      class="text-center"
    >
      <h1
        class="text-4xl font-bold text-gray-900 mb-4"
      >
        Oops!
      </h1>
      <p
        class="text-lg text-gray-600 mb-8"
      >
        Sorry, an unexpected error has occurred.
      </p>
      <p
        class="text-gray-500"
      >
        <i>
          Unknown error
        </i>
      </p>
    </div>
  </div>
</div>
`;
